feat: Implement AsyncReadRent/WriteRent for std in-memory types

VanjaRo opened the pull request:
Addresses [#310](https://github.com/bytedance/monoio/issues/310).

**Changes based on the commit:**

*   **`AsyncReadRent`:**
    *   Implemented for `&mut [u8]`.
*   **`AsyncWriteRent`:**
    *   Refactored for `Vec<u8>`.
    *   Implemented for `Cursor<&mut Vec<u8>>`.
    *   Implemented for `Cursor<&mut [u8]>`.
    *   Implemented for `Cursor<Box<[u8]>>`.
    *   Implemented for `Cursor<Vec<u8>>`.
    *   Implemented for `Box<T: ?Sized + AsyncWriteRent + Unpin>`.
*   **Helpers:** Added helper functions (`iovecs_to_slices`, `wsabufs_to_slices`, `cursor_writev_logic`) to support `writev` implementations, particularly for `Cursor`.
*   **Tests:** Added/verified test coverage for the new implementations:
    *   `tests/async_read_rent_mut_slice.rs`: Covers `AsyncReadRent for &mut [u8]`.
    *   `tests/async_write_rent_cursor.rs`: Covers `AsyncWriteRent for Cursor<T>` types.
    *   `tests/async_write_rent_boxed.rs`: Covers `AsyncWriteRent for Box<T>`.

VanjaRo committed:
feat: implement AsyncReadRent and AsyncWriteRent for mutable slices and Cursor types

- Added `AsyncReadRent` implementation for `&mut [u8]`, including `read` and `readv` methods.
- Introduced `AsyncWriteRent` implementation for `Cursor` types.
- Added tests for reading and writing with mutable slices and Cursor to ensure functionality and correctness.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/bytedance/monoio?pullRequest=344) <br/>All committers have signed the CLA.

Lzzzzzt commented on monoio/src/io/async_read_rent.rs:
```diff
@@ -128,6 +128,66 @@ impl AsyncReadRent for &[u8] {
+impl AsyncReadRent for &mut [u8] {
+    fn read<T: IoBufMut>(&mut self, mut buf: T) -> impl Future<Output = BufResult<usize, T>> {
+        let amt = std::cmp::min(self.len(), buf.bytes_total());
+        let (a_ptr, b_ptr, b_len) = {
```
Could you let these variables be more readable

VanjaRo replied:
Was operating within namings that were already present within a file. Made the variables more meaningful.

Lzzzzzt commented on monoio/src/io/async_write_rent.rs:
```diff
@@ -116,6 +118,57 @@ impl<A: ?Sized + AsyncWriteRent> AsyncWriteRent for &mut A {
 }
 
+#[cfg(unix)]
+fn iovecs_to_slices(iov_ptr: *const libc::iovec, iov_len: usize) -> Vec<&'static [u8]> {
```
The heap allocation here seems not necesssary? Since the Cursor's implement of `Write` is basically using the for-loop, we could also do that the avoid this allocation.

VanjaRo replied:
Refactored this logic, now operating within raw-constructed slices.

VanjaRo committed:
refactor: refactored namings for read and optimized write implementations

VanjaRo commented:
Also, unified the Cursor implementation to the "generic" one, as the specifications for each individual type seems to bring no benefit for the codebase.

VanjaRo committed:
fix: fmt fix

Lzzzzzt commented on monoio/src/io/async_write_rent.rs:
```diff
@@ -116,6 +119,69 @@ impl<A: ?Sized + AsyncWriteRent> AsyncWriteRent for &mut A {
+    #[cfg(windows)]
+    {
+        // SAFETY: IoSlice<'_> is repr(transparent) over WSABUF
+        let bufs = unsafe {
```
This leaves `bufs` uninitialised

VanjaRo replied:
Done

VanjaRo committed:
fix: redundant `let` in windows

VanjaRo committed:
fix: linter

VanjaRo commented:
Can't reproduce/debug the failing CI locally, **need help**. 
Tried multiple architectures –– every setup finishes with `ok`
Seems that `legacy_accept_read_write` and `legacy_shutdown` tests issues are unrelated to my PR

VanjaRo committed:
fix: windows clippy

VanjaRo committed:
fix: update pin macro examples and handle unsupported errors for UNIX domain sockets

- Added missing `use monoio::pin;` in examples to ensure correct usage.
- Updated error handling for `set_reuse_port` and `set_reuse_address` to ignore unsupported errors on UNIX domain sockets.

Lzzzzzt commented on monoio/src/net/unix/listener.rs:
```diff
@@ -38,7 +38,15 @@ impl UnixListener {
         let addr = socket2::SockAddr::unix(path)?;
 
         if config.reuse_port {
-            sys_listener.set_reuse_port(true)?;
```
Could you explain why this error should be ignored

Lzzzzzt replied:
Other parts look good to me.

Lzzzzzt replied:
> Could you explain why this error should be ignored

Is this cause the CI test failed before?

VanjaRo replied:
> > Could you explain why this error should be ignored
> 
> Is this cause the CI test failed before?

Yes, I triggered the master branch CI within cloned repo –– all `linux` setups stumble upon this missing `SO_REUSEPORT` syscall.
You can look at this action from my cloned version [action failure](https://github.com/VanjaRo/monoio/actions/runs/14973370110/job/42059392758) and look at the parent commits of it.

VanjaRo replied:
As detailed in the GitHub issue I found https://github.com/amazonlinux/amazon-linux-2023/issues/901, SO_REUSEPORT was restricted to INET sockets (like TCP/IP) in the Linux kernel (see kernel commit 5b0af621c3f6). This was done to fix a bug related to another protocol that did not support SO_REUSEPORT.
Previously, SO_REUSEPORT could be set on any socket family, but it only had an actual effect on INET sockets. The change means that attempting to set SO_REUSEPORT on an AF_UNIX socket now returns an ENOTSUP ("Operation not supported") error. This is not a bug in the distributions or the CI setup, but rather the kernel behaving as intended. I guess some refactoring to socket logic is needed here.

Lzzzzzt replied:
> As detailed in the GitHub issue I found [amazonlinux/amazon-linux-2023#901](https://github.com/amazonlinux/amazon-linux-2023/issues/901), SO_REUSEPORT was restricted to INET sockets (like TCP/IP) in the Linux kernel (see kernel commit 5b0af621c3f6). This was done to fix a bug related to another protocol that did not support SO_REUSEPORT. Previously, SO_REUSEPORT could be set on any socket family, but it only had an actual effect on INET sockets. The change means that attempting to set SO_REUSEPORT on an AF_UNIX socket now returns an ENOTSUP ("Operation not supported") error. This is not a bug in the distributions or the CI setup, but rather the kernel behaving as intended. I guess some refactoring to socket logic is needed here.

Ok, got it.

Lzzzzzt replied:
> As detailed in the GitHub issue I found [amazonlinux/amazon-linux-2023#901](https://github.com/amazonlinux/amazon-linux-2023/issues/901), SO_REUSEPORT was restricted to INET sockets (like TCP/IP) in the Linux kernel (see kernel commit 5b0af621c3f6). This was done to fix a bug related to another protocol that did not support SO_REUSEPORT. Previously, SO_REUSEPORT could be set on any socket family, but it only had an actual effect on INET sockets. The change means that attempting to set SO_REUSEPORT on an AF_UNIX socket now returns an ENOTSUP ("Operation not supported") error. This is not a bug in the distributions or the CI setup, but rather the kernel behaving as intended. I guess some refactoring to socket logic is needed here.

cc @ihciah

Lzzzzzt commented:
> Can't reproduce/debug the failing CI locally, **need help**. Tried multiple architectures –– every setup finishes with `ok` Seems that `legacy_accept_read_write` and `legacy_shutdown` tests issues are unrelated to my PR

Are you use `cross` to run the test or on your own PC? May be the error is related to the `cross`.

VanjaRo commented:
> > Can't reproduce/debug the failing CI locally, **need help**. Tried multiple architectures –– every setup finishes with `ok` Seems that `legacy_accept_read_write` and `legacy_shutdown` tests issues are unrelated to my PR
> 
> Are you use `cross` to run the test or on your own PC? May be the error is related to the `cross`.

I'm running all tests on the GitHub CI env to eliminate any compatibility issues.

Lzzzzzt committed:
Merge branch 'master' into async-rent

Lzzzzzt commented:
Thank you so much!
